packages/flutter/lib/src/widgets/text_selection.dart

```diff
@@ -1904,7 +1934,29 @@ class TextSelectionGestureDetectorBuilder {
             case PointerDeviceKind.touch:
             case PointerDeviceKind.unknown:
               // On iOS/iPadOS a touch tap places the cursor at the edge of the word.
-              renderEditable.selectWordEdge(cause: SelectionChangedCause.tap);
+              final TextSelection previousSelection = editableText.textEditingValue.selection;
+              // If the `previousSelection` is collapsed, the tap was on the selection (including the bounds), and the
+              // editable is focused, then the toolbar is toggled.
+              //
+              // If the `previousSelection` is not collapsed, the tap was on the selection (excluding the bounds, we exclude the
+              // the bounds so we can still tap on the selection edges to select the word edge), and the editable is focused, then
+              // the toolbar is toggled.
+              //
+              // If the tap was neither on the selection exclusively or inclusively then we select the word edge. After
+              // selecting the word edge, if our selection remains the same, then we toggle the toolbar. If the selection changes
+              // then we hide the toolbar.
+              if (((_tapWasOnSelectionExclusive(details.globalPosition) && !previousSelection.isCollapsed)
```

Comment:
This is a bit hard to read. Could you simplify this a bit?

Reply:
Also you probably don't need `_tapWasOnSelectionInclusive`? if `previousSelection == editableText.textEditingValue.selection` then we'll toggle the toolbar anyway?

Reply:
I think there is still the case where the cursor is in the middle of a word `wor|d`, and we tap it.

If we remove the `inclusive` check, then the `exclusive` one will not catch it, and the cursor will move to the word edge.

The expected behavior would be to toggle the toolbar when we tap the cursor.

Reply:
```dart
if () {
   // If the tap was on the selection (different meaning for collapsed vs uncollapsed selection) and the editable was focused, then toggle the toolbar.
} else if () {
   // Tap was not on the selection at all.
   // We should still toggle the toolbar, if the selection remains the same after the word edge is selected.
   // If the selection changes then we hide the toolbar.
}
```

Reply:
I updated the logic a little, getting rid of the `else if`.

Reply:
> Tap was not on the selection at all.
> if the selection remains the same after the word edge is selected.
can the two both be true?

Reply:
Yeah, this can happen in the following example

'This is an example'

tap the position where 'p' is.
-> Cursor will move to closest word edge `example|`

tap the position where 'p' is again.
-> Cursor will move to closest word edge 'example|'
-> The previousSelection is the same as the new selection after the closest word edge is selected, so the toolbar toggles

Reply:
For the collapsed tap check you'll have to take text affinity into account I think. For instance if the cursor is at the end of a soft-wrapped line and you tap the start of the next line, you'll get a TextPosition that has the same offset as the caret location.

Reply:
Hmm but if the cursor is located in a place when the affinity doesn't matter then that will be too strict. I guess we can fix that later.

Reply:
Also could you explain why `RenderEditable.hasFocus` need to be taken into account?

Reply:
Yeah I’ll keep a lookout of affinity issues.

On native if the renderEditable doesn’t have focus we always select the closest word edge. And the toolbar is not toggled unless the editable has focus.

Reply:
Could you check if you tap the latter half of the character before the caret? I think that will give you (x-1, upstream) instead of (x, downstream) so `wasTapOnSelection` will return false? But that should also toggle the toolbar right?

Reply:
Good point, If i do the later half of the character before the character the selection is upstream. If I then follow that up with a tap on the former half of the character after the caret, then that selection is downstream. Then because of `previousSelection == editableText.textEditingValue.selection`, the toolbar is not toggled in this case.

Reply:
I think in the case where we select the word edge the affinity does not matter since a word will not be wrapped between two lines (will either fall on first or second line).

In the case where we check for `wasTapOnSelectionInclusive` we do care about the affinity because we want to be able to select the text at beginning of the next line. So we want it to return false when the affinity is different.

Reply:
There might be a trade off we have to make here since we cannot differentiate why the TextAffinity changed.

We want to toggle the toolbar when
- We call `selectWordEdge` and the selection after the call is the same as the previous selection (including TextAffinity).
- We call `selectWordEdge` and the selection after the call is the same as the previous selection (excluding TextAffinity), this can happen when we tap 

We want to hide the toolbar when 
- We call `selectWordEdge` and the selection after the call is the same as the previous selection (excluding TextAffinity). This can happen when there is a multi-line text field, there is word wrap and we want to tap on the boundary. Let's say the cursor is at the end of the first line. Tapping on the first line boundary should toggle the toolbar, tapping at the beginning of the second line should hide the toolbar.

We want differentiate between whether `TextAffinity` changed due to tapping on a new line but same selection offset, or if it occurred from tapping on the same line with the same selection offset.